llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -292,6 +292,70 @@ InstructionCost VPRecipeBase::computeCost(ElementCount VF,
   llvm_unreachable("subclasses should implement computeCost");
 }
 
+InstructionCost
+VPPartialReductionRecipe::computeCost(ElementCount VF,
+                                      VPCostContext &Ctx) const {
+  std::optional<unsigned> Opcode = std::nullopt;
+  VPRecipeBase *BinOpR = getOperand(0)->getDefiningRecipe();
+  if (auto *WidenR = dyn_cast<VPWidenRecipe>(BinOpR))
+    Opcode = std::make_optional(WidenR->getOpcode());
+
+  VPRecipeBase *ExtAR = BinOpR->getOperand(0)->getDefiningRecipe();
+  VPRecipeBase *ExtBR = BinOpR->getOperand(1)->getDefiningRecipe();
+
+  auto *PhiType = Ctx.Types.inferScalarType(getOperand(1));
+  auto *ExtTy = ExtAR ? Ctx.Types.inferScalarType(ExtAR->getOperand(0))
```

Comment:
IICU this now assumes that ExtAR will always be either a live-in or an extend. Is that always the case or should this actually check if it is an extend?

Reply:
There  is the check on line 325 with `GetExtendKind(ExtAR)`, if that's what you mean?

Reply:
Yep, IIUC `ExtTy` should be the input type (before extending); if that's the case, might be good to align the name here with the one passed to TTI.

 If `AR` isnt't an extend, but `BR` is, what is the expected type to be passed to the TTI helper? What if `AR` and `BR` are extends, but with different input widths (eg `zext i8 to i32` and `zext i16 to i32`?

Reply:
Renaming it to `InputType` sounds like a good idea to me.

At the moment both A and B have to be extends from and to the same type but I can see the argument for passing both input types to the target and letting it decide.